kore/src/Kore/Unification/NewUnifier.hs

```diff
@@ -844,8 +837,9 @@ unifyTerms' rootSort sideCondition origVars vars ((first, second) : rest) bindin
             } =
             let normalElementsWithVariables = map getSetElement elementsWithVariables
                 normalConcreteElements = map normalizeConcreteSetElement $ HashMap.toList concreteElements
-                variables = MultiSet.fromList $ map elemVar $ filter isVar opaque
-                functions = MultiSet.fromList $ filter (not . isVar) opaque
+                (!variables, !functions) = (MultiSet.fromList vrs, MultiSet.fromList fns)
+                  where
+                    (fns, vrs) = partitionEithers (map elemVar opaque)
                 elements = Set.map (reconstructSetElement element) $ Set.fromList $ normalElementsWithVariables ++ normalConcreteElements
```

Comment:
What is going on in this line? Are there potential duplicates in the list that `fromList` is removing? `Set.map` is on the expensive side, so unless doing so could lead to expensive recomputations it would likely be better to map over the list before converting it to a set.

Reply:
I also see a suspicious amount of code duplication between these lines and the ones just above in the diff. Should that be factored out, or is that just a coincidence?

Reply:
I see no reason to expect there to be enough duplicates to matter for performance. Mapping before converting to a set ought to be fine unless you notice a measurable slowdown.

Reply:
As far as the set/map code goes, I tried to make as much of the code path between maps and sets be shared in common as possible. If you think this can be improved further, go right ahead.